Initial precompiled shaders implementation

inner-daemons opened the pull request:
**Connections**
Works towards #3103
Depends on #7831 

Also, as this is the start of a big change, I'd like to ping @cwfitzgerald to at least make sure I haven't gone off in the wrong direction.

**Description**
This adds a `CreateShaderModuleDescriptorPassthrough::Generic` enum variant which contains code for multiple types of shader source,allows creating passthrough shaders without writing backend specific code (if on metal pass MSL, etc). This variant also includes an optional reflection thing. For now, I don't know exactly what reflection info is needed, but if possible, we should make sure this can live in `wgpu-types` or `wgpu-core` to avoid dependency on `naga`. It seems the best model for this is the `wgpu_core::validation::Interface`, we might just have to replace some of the naga types.

Using this requires enabling the EXPERIMENTAL_PRECOMPILED_SHADERS feature. This feature is only supported on DX12, Vulkan, and Metal. Logic on these backends is otherwise identical to the respective specific passthrough methods.

Nothing is currently done with the reflection info.

An overview of my approach can be found in [this comment](https://github.com/gfx-rs/wgpu/issues/3103#issuecomment-2833058367). If this process takes longer than expected we can make a tracking issue. For now I will be posting updates by editing that comment and referencing it in PRs.

**Testing**
No testing yet. However, the code seems somewhat small and robust.

Squash pls

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `taplo format`.
- [x] Run `cargo clippy --tests`. If applicable, add:
  - [x] `--target wasm32-unknown-unknown`
- [x] Run `cargo xtask test` to run tests.
- [x] If this contains user-facing changes, add a `CHANGELOG.md` entry. <!-- See instructions at the top of `CHANGELOG.md`. -->

inner-daemons committed:
Unsure but these were prior changes

inner-daemons committed:
Merge branch 'dxil-passthrough' of https://github.com/supamaggie70incorporated/wgpu into dxil-passthrough

inner-daemons committed:
Fixed feature thing

inner-daemons committed:
Mostly finished dxil passthrough

inner-daemons committed:
Attempted to add HLSL passthrough

inner-daemons committed:
Tried to fix compiles

inner-daemons committed:
Tried to fix more clippy lints

inner-daemons committed:
Added initial feature, going back to fix issues with DXIL rn

inner-daemons committed:
Tried to fix some potential issues

inner-daemons committed:
Merge branch 'dxil-passthrough' into precompiled-shaders

inner-daemons committed:
Initial work

inner-daemons committed:
Tried to fix metal compiles

inner-daemons committed:
Fixed warning

inner-daemons committed:
Added my ideas for changes

inner-daemons committed:
Tried to expose precompiled shaders to metal

cwfitzgerald commented:
Hmm, I think in general I like the ideas put forward in your plan you wrote up, that makes sense to me. As for the code changes themselves, I don't think that we need another variant internally - this feels redundant when all the other variants exist. Maybe what would need to be added is an optional reflection info to each of the variants?

inner-daemons commented:
@cwfitzgerald The purpose of this variant is so that the user can pass a single shader without considering backend-specific code. This way we can also later on add a macro that creates this struct, filling fields for all the backends that are supported, and it doesn't need to be passed a `Backend` parameter. I can change the approach if you want, but it might make everything a little less clean from the user's point of view.

inner-daemons committed:
Merge branch 'trunk' into precompiled-shaders

cwfitzgerald commented:
Yeah, I agree that it makes sense _in `wgpu's` api_ but wgpu-core and wgpu-hal should be able to use the original ones?

inner-daemons commented:
> Yeah, I agree that it makes sense _in `wgpu's` api_ but wgpu-core and wgpu-hal should be able to use the original ones?

True, didn't think about that to be honest. I'll rewrite that part of this PR

inner-daemons committed:
Merge branch 'trunk' into precompiled-shaders

inner-daemons committed:
How do these things even happen?

inner-daemons committed:
Merge branch 'trunk' into precompiled-shaders

inner-daemons committed:
Merge branch 'trunk' into precompiled-shaders

inner-daemons committed:
More work, adapted with some of Cw's advice

inner-daemons committed:
Tried to fix some compile errors. Now build!!!

inner-daemons committed:
Merge branch 'trunk' into precompiled-shaders

inner-daemons commented on wgpu-core/src/pipeline.rs:
```diff
@@ -147,6 +149,8 @@ impl WebGpuError for CreateShaderModuleError {
             #[cfg(feature = "spirv")]
             Self::ParsingSpirV(..) => return ErrorType::Validation,
+            // TODO: ensure this is the correct error type
+            Self::NotCompiledForBackend => return ErrorType::Validation,
```
@cwfitzgerald Can you confirm this is correct?

cwfitzgerald replied:
Yes I believe it is

inner-daemons commented on wgpu-types/src/features.rs:
```diff
@@ -1267,6 +1267,16 @@ bitflags_array! {
+        /// - OpenGL ES
+        ///
+        /// This is a native only feature.
+        const GLSL_SHADER_PASSTHROUGH = 1 << 55;
```
Should I move this into a separate PR? It hasn't been implemented yet

cwfitzgerald replied:
I think this is fine.

inner-daemons committed:
Merge branch 'trunk' into precompiled-shaders

cwfitzgerald commented on wgpu-types/src/features.rs:
```diff
@@ -1254,6 +1254,29 @@ bitflags_array! {
+        ///
+        /// GLSL data is not parsed or interpreted in any way
+        ///
+        /// Supported platforms:
```
```suggestion
        /// This features is not implemented yet.
        /// 
        /// Potential supported platforms:
```

inner-daemons committed:
Small change

inner-daemons committed:
Update wgpu-types/src/features.rs

Co-authored-by: Connor Fitzgerald <connorwadefitzgerald@gmail.com>

inner-daemons committed:
Merge branch 'trunk' into precompiled-shaders

inner-daemons committed:
Deleted the precompiled shader spec

cwfitzgerald reviewed (changes requested):
Sitting down and thinking about this api, I feel there's redundancy here by having both all the individual variants and the combined variant. I think we should have either use the ShaderModuleDescriptorGeneric as-is as the passthrough shader descriptor, or we should ShaderModuleDescriptorGeneric but with individual members replaced with a single enum, which the user of that can then chose. Whatever is generating it can check which backend the device is using.

inner-daemons commented:
@cwfitzgerald I think I prefer the first option here, at least for the user facing option. But I do agree that there is some redundancy here as is. Are you willing to make a decision on which approach to use yet? I'd like to commit to the first and just try to get this pushed through soon (obviously not for 26.0, but I'd like to start working on more aspects on precompiled shaders).

inner-daemons committed:
Merge branch 'trunk' into precompiled-shaders

cwfitzgerald commented:
@SupaMaggie70Incorporated Alright, lets go with option 1 then! We can always iterate throughout the v26->27 cycle.

inner-daemons commented:
@cwfitzgerald I'm guessing I should keep the existing passthrough features?

cwfitzgerald commented:
Hmm. Traditionally we have said you should be able to derive all the information you need from the adapters. However, this isn't true for all the interop or passthrough features. I think that we should just have a single feature for passthrough in general, then you can look at the backend of the adapter.

inner-daemons committed:
Pushing now to get hal errors on other backends

inner-daemons committed:
Made some slight fixes, CI will now report more hal errors

inner-daemons committed:
Tried to write passthrough for webgpu backend

inner-daemons committed:
Merge branch 'trunk' into precompiled-shaders

inner-daemons commented:
@cwfitzgerald Should be good now :) I also added passthrough for WGSL. Its only missing GLSL at this point, so OpenGL doesn't expose the feature. There is no testing for any of this as far as I'm aware, let me know if I should work on that or postpone it to another PR.

inner-daemons committed:
Merge branch 'trunk' into precompiled-shaders

inner-daemons committed:
Merge branch 'trunk' into precompiled-shaders

inner-daemons committed:
Merge branch 'trunk' into precompiled-shaders

cwfitzgerald commented on wgpu-hal/src/dx12/device.rs:
```diff
@@ -1701,10 +1701,13 @@ impl crate::Device for super::Device {
             }),
             crate::ShaderInput::SpirV(_) => {
-                panic!("SPIRV_SHADER_PASSTHROUGH is not enabled for this backend")
+                panic!("Features::SPIRV_SHADER_PASSTHROUGH is not enabled for this backend")
             }
             crate::ShaderInput::Msl { .. } => {
-                panic!("MSL_SHADER_PASSTHROUGH is not enabled for this backend")
+                panic!("Features::MSL_SHADER_PASSTHROUGH is not enabled for this backend")
+            }
+            crate::ShaderInput::Glsl { .. } => {
+                panic!("Features::GLSL_SHADER_PASSTHROUGH is not enabled for this backend")
             }
```
These need to be updated to use the new feature name

cwfitzgerald commented on wgpu-hal/src/metal/device.rs:
```diff
@@ -1037,11 +1037,14 @@ impl crate::Device for super::Device {
                 })
             }
             crate::ShaderInput::SpirV(_) => {
-                panic!("SPIRV_SHADER_PASSTHROUGH is not enabled for this backend")
+                panic!("Features::SPIRV_SHADER_PASSTHROUGH is not enabled for this backend")
             }
             crate::ShaderInput::Dxil { .. } | crate::ShaderInput::Hlsl { .. } => {
                 panic!("`Features::HLSL_DXIL_SHADER_PASSTHROUGH` is not enabled for this backend")
             }
+            crate::ShaderInput::Glsl { .. } => {
+                panic!("Features::GLSL_SHADER_PASSTHROUGH is not enabled for this backend")
+            }
```
Same here

cwfitzgerald commented on wgpu-hal/src/vulkan/device.rs:
```diff
@@ -1914,9 +1914,12 @@ impl crate::Device for super::Device {
                 panic!("MSL_SHADER_PASSTHROUGH is not enabled for this backend")
             }
             crate::ShaderInput::Dxil { .. } | crate::ShaderInput::Hlsl { .. } => {
-                panic!("`Features::HLSL_DXIL_SHADER_PASSTHROUGH` is not enabled")
+                panic!("`Features::HLSL_DXIL_SHADER_PASSTHROUGH` is not enabled for this backend")
             }
             crate::ShaderInput::SpirV(spv) => Cow::Borrowed(spv),
+            crate::ShaderInput::Glsl { .. } => {
+                panic!("Features::GLSL_SHADER_PASSTHROUGH is not enabled for this backend")
+            }
```
And here

cwfitzgerald commented on wgpu-types/src/lib.rs:
```diff
@@ -7757,155 +7757,85 @@ pub enum DeviceLostReason {
-            CreateShaderModuleDescriptorPassthrough::Msl(inner) => inner.source.as_bytes(),
-            CreateShaderModuleDescriptorPassthrough::Dxil(inner) => inner.source,
-            CreateShaderModuleDescriptorPassthrough::Hlsl(inner) => inner.source.as_bytes(),
+        if let Some(spirv) = &self.spirv {
+            bytemuck::cast_slice(spirv)
+        } else if let Some(msl) = &self.msl {
+            msl.as_bytes()
+        } else if let Some(dxil) = &self.dxil {
+            dxil
+        } else {
+            panic!("No binary data provided to `ShaderModuleDescriptorGeneric`")
```
I don't think we can have a single data file anymore, we need to have a single file for each type. Additionally we need to output the hlsl etc. With tracing we need to be able to replicate all the descriptors.

inner-daemons replied:
@cwfitzgerald I wonder if for tracing we should only use the source that is actually used. And yeah good catch, this was just cobbled together so it would still build and have "some" level of tracing

cwfitzgerald replied:
I think the intent was that traces should be able to work cross api, but I don't know if that is currently true. The trace infrastructure is broken and pretty neglected.

inner-daemons replied:
I've attempted to address this. Let me know if you think my approach is workable. I haven't done any testing or thought too hard about it, but as you mentioned tracing is in a very sorry state right now.

cwfitzgerald replied:
Alright, I'm not too worried about it as we can't even  record traces right now!

cwfitzgerald reviewed (changes requested):
I added some comments on the code, I want to do some docs adjustments before this lands, but github won't let me do it inline, so I'll push a commit when I have a minute

inner-daemons committed:
Merge branch 'trunk' into precompiled-shaders

inner-daemons committed:
Various changes as requested from review

inner-daemons committed:
Fixed issues with trace

inner-daemons committed:
Merge branch 'trunk' into precompiled-shaders

inner-daemons committed:
Merge branch 'trunk' into precompiled-shaders

inner-daemons commented:
> I added some comments on the code, I want to do some docs adjustments before this lands, but github won't let me do it inline, so I'll push a commit when I have a minute

All your comments should be addressed :)

inner-daemons committed:
Merge branch 'trunk' into precompiled-shaders

inner-daemons committed:
Merge branch 'trunk' into precompiled-shaders

inner-daemons committed:
Fixed accidental merge change

inner-daemons committed:
Merge branch 'trunk' into precompiled-shaders

jimblandy commented:
@cwfitzgerald When you're back - this has been waiting for a while.

DJMcNab commented on player/src/lib.rs:
```diff
@@ -315,6 +315,94 @@ impl GlobalPlay for wgc::global::Global {
+                        let mut file = fs::File::open(dir.join(a)).unwrap();
+                        let word_count = file.metadata().unwrap().len() as usize / 4;
+                        let mut vec = Vec::<u32>::with_capacity(word_count);
+                        unsafe {
+                            file.read_exact(std::slice::from_raw_parts_mut(
+                                vec.as_mut_ptr() as *mut u8,
+                                vec.len() * 4,
+                            ))
+                            .unwrap();
+                            vec.set_len(word_count);
+                        }
```
Maybe I've missed some previous GitHub comment in the PR here, but this at least needs a safety comment.

Are you sure that this isn't just busted anyway? `vec.len() * 4` will be 0 * 4, which is 0, and then we set the length to a non-zero value. Also, even if this code did work, it would be insta-UB, because we're creating a slice with uninitialised contents.

I definitely get the impulse to not read into a u8 vector, then copy to a u32 vector. That's also definitely wrong.

(Also, if the size isn't divisible by 4, this seems likely to be wrong)

inner-daemons replied:
I think I need my unsafe privileges revoked. Never goes well!

DJMcNab commented on wgpu-types/src/features.rs:
```diff
@@ -1233,15 +1210,18 @@ bitflags_array! {
-        /// Supported platforms:
+        /// Supported platforms
+        /// - Vulkan
         /// - DX12
```
Is Metal not supported here? I see that we previously supported the `MSL_SHADER_PASSTHROUGH` feature.

DJMcNab commented on wgpu-types/src/features.rs:
```diff
@@ -1233,15 +1210,18 @@ bitflags_array! {
         const EXTENDED_ACCELERATION_STRUCTURE_VERTEX_FORMATS = 1 << 50;
 
-        /// Enables creating shader modules from DirectX HLSL or DXIL shaders (unsafe)
+        /// Enables creating shaders from passthrough with reflection info (unsafe)
```
Can we link to the relevant API (even though it will need to be hardcoded to docs.rs). E.g. how we linked to `include_spirv_raw`.

inner-daemons replied:
Is there a good reason why we should do that instead of e.g. just giving the function name? Linking directly to docs.rs seems like it would come with some potential issues, like linking to more recent or non-existent docs as the API changes (or before it is part of a release)

DJMcNab replied:
It's not super important. Again, I was mostly seeing that it was incongruent with the old API, which did link to [`include_spirv_raw`](https://docs.rs/wgpu/latest/wgpu/macro.include_spirv_raw.html) - though for some reason, calling it `make_spirv_raw`. See also [`Self::PUSH_CONSTANTS`](https://docs.rs/wgpu/latest/wgpu/struct.Features.html#associatedconstant.PUSH_CONSTANTS), as an example.

That is, these concerns have already been litigated, I believe.

cwfitzgerald replied:
> Linking directly to docs.rs seems like it would come with some potential issues, like linking to more recent or non-existent docs as the API changes (or before it is part of a release)

I think we should be linking, it's a bit unfortunate we can't validate it, but the way our docs are defined have hamstrung us a bit.

DJMcNab commented on wgpu-types/src/lib.rs:
```diff
@@ -7838,155 +7838,87 @@ pub enum DeviceLostReason {
-/// This type is unique to the Rust API of `wgpu`. In the WebGPU specification,
-/// only WGSL source code strings are accepted.
+/// Descriptor for a shader module given by any of several sources, with optional reflection information.
+/// All shader types that may be used by the backend must be `Some`, otherwise usage is undefined behavior
```
This strikes me as odd. We panic in the cases where this is none, which isn't really undefined behaviour.

DJMcNab commented on wgpu-types/src/lib.rs:
```diff
@@ -7838,155 +7838,87 @@ pub enum DeviceLostReason {
+    pub label: L,
+    /// Number of workgroups in each dimension x, y and z. Unused for Spir-V.
+    pub num_workgroups: (u32, u32, u32),
+    /// Optional reflection information
+    pub reflection: Option<ShaderModuleReflection>,
```
We should better document when it is valid to leave this out/what is it used for/how do you get it?

Also missing full stop at end of doc comment.

inner-daemons replied:
Reflection information is always optional. I left this as a field so we could work from it later on. Right now the struct is actually empty. There isn't a way to actually construct it yet. I can remove it if you think thats best for now, just wanted to get out a rough sketch of the API.

cwfitzgerald replied:
I think removing it is fine. We can add it in the future.

DJMcNab reviewed (commented):
I'm really looking forward to this! Just some things which popped up!

inner-daemons commented:
@DJMcNab Thanks for the review! Upon further inspection it seems my PR was not super well thought through or checked, so I'll have another look over everything myself in addition to addressing your comments.

inner-daemons committed:
Merge branch 'trunk' into precompiled-shaders

inner-daemons committed:
Tried to fix the tests and whatnot

inner-daemons committed:
More small tweaks

inner-daemons committed:
A few more tweaks

inner-daemons committed:
Tried to fix spirv issue (I'm blind to vulkan compile rn)

inner-daemons committed:
Ok thats embarassing

inner-daemons committed:
Final pass (single word changed in doc comment)

inner-daemons commented:
@DJMcNab I've just given this PR a final look and it looks good. Some of your comments I left open as I wasn't sure how to handle them. Let me know what you think, you get the final say of course!

Also pinging @cwfitzgerald because this hopefully can be merged soon

Another note: the only backend *not* currently supported is OpenGL. Once we get that done, maybe we should consider removing this as a feature? I'm not sure how that would interact with e.g. custom backends so I'll wait until later to inquire about that.

DJMcNab commented:
I'm not planning to review this again - it needs to be reviewed by someone with write access, and I was providing a "first pass" review to help speed this along, but I can't give a fully "in-context" review.

Thanks for actioning the review feedback so well! I've found it easy to submit PRs which don't hold up when given a second reading myself as well - that's what review is for, after all!

cwfitzgerald commented on player/src/lib.rs:
```diff
@@ -315,6 +315,90 @@ impl GlobalPlay for wgc::global::Global {
+                    if a.ends_with(".spv") {
+                        let data = fs::read(dir.join(a)).unwrap();
+                        assert!(data.len() % 4 == 0);
+                        // Mom: we have bytemuck at home
+                        let vec: Vec<u32> = data
+                            .chunks_exact(4)
+                            .map(|chunk| u32::from_ne_bytes(chunk.try_into().unwrap()))
+                            .collect();
```
Lmaooo just take a bytemuck dep

DJMcNab replied:
As far as I know, this isn't something bytemuck can help with, because u32 and u8 have different alignments. I think you can have a happy-path if the `Vec<u8>` happened to be allocated with a 4 byte alignment, but even still, you can't guarantee it.

This is an area which just really sucks; I don't think there's any way to actually spell the semantics you want here in stable Rust. We've ran into the same thing with preparing buffers in Vello.

cwfitzgerald replied:
There is actually! Thanks to this lovely function https://docs.rs/bytemuck/latest/bytemuck/allocation/fn.pod_collect_to_vec.html

cwfitzgerald commented on wgpu-core/src/device/resource.rs:
```diff
@@ -2125,39 +2125,59 @@ impl Device {
+
+        // TODO: when we get to use if-let chains, this will be a little nicer!
+
+        log::info!("Backend: {}", self.backend());
```
Forgotalog

cwfitzgerald commented on wgpu-types/src/lib.rs:
```diff
@@ -7995,155 +7995,108 @@ pub enum DeviceLostReason {
+    pub hlsl: Option<Cow<'a, str>>,
+    /// Shader GLSL source (currently unused).
+    pub glsl: Option<Cow<'a, str>>,
+    /// Shader WGSL source (currently unused).
```
This isn't true anymore.
```suggestion
    /// Shader WGSL source.
```

cwfitzgerald commented on wgpu-types/src/lib.rs:
```diff
@@ -7995,155 +7995,108 @@ pub enum DeviceLostReason {
-///
-/// This type is unique to the Rust API of `wgpu`. In the WebGPU specification,
-/// only WGSL source code strings are accepted.
+/// Descriptor for a shader module given by any of several sources, with optional reflection information.
```
```suggestion
/// Descriptor for a shader module given by any of several sources, with optional reflection information.
/// These shaders are passed through directly to the underlying api.
```

cwfitzgerald commented on CHANGELOG.md:
```diff
@@ -53,6 +53,9 @@ We have merged the acceleration structure feature into the `RayQuery` feature. T
 By @Vecvec in [#7913](https://github.com/gfx-rs/wgpu/pull/7913).
 
+### General
+- Added `Features::EXPERIMENTAL_PRECOMPILED_SHADERS`, replacing existing passthrough types with a unified `CreateShaderModuleDescriptorPassthrough` which allows passing multiple shader codes for different backends. By @SupaMaggie70Incorporated in [#7834](https://github.com/gfx-rs/wgpu/pull/7834)
```
This should be a major changes with a small things showing off the diff and the new API

inner-daemons replied:
I made a new changelog entry, let me know if you want more changes to it

cwfitzgerald reviewed (changes requested):
A few small comments, but this is exactly what I was looking for!

inner-daemons committed:
Merge branch 'trunk' into precompiled-shaders

inner-daemons committed:
Updated changelog

inner-daemons committed:
Apply suggestions from code review

Co-authored-by: Connor Fitzgerald <connorwadefitzgerald@gmail.com>

inner-daemons committed:
Addressed more comments

inner-daemons committed:
Removed references to optional reflection information

inner-daemons commented:
Alright @cwfitzgerald, I think I've addressed most of the recent changes (and also updated the PR to latest trunk). If there's anything more let me know.

inner-daemons committed:
Merge branch 'trunk' into precompiled-shaders

inner-daemons committed:
Merge branch 'trunk' into precompiled-shaders

cwfitzgerald reviewed (approved):
Lets do it!

Could you file an issue (or make a PR) adding tests for passing through to all the backends
